pages/en/games/handling-input-changes.md

```diff
@@ -0,0 +1,357 @@
+
+### Prioritized state machine based on received events
+
+One approach is to keep track of the current “active input state” - representing the input prompts currently shown on-screen to the user - by keeping track of actual input events being received by the app and transitioning between the states using prioritized logic.
+
+#### Prioritize
+
+Why prioritize input states? Users interact with their devices in all sorts of ways and your app should support their choice. For example, if a user chooses to use the touchscreen and a bluetooth mouse at the same time, that should be supported. But which on-screen input prompts and controls should you display? Mouse or touch?
+
+Defining each set of prompts as an “input state” and then prioritizing them can help decide this in a consistent way.
+
+#### Received input events determine the state
+
+Why only act on received input events? You might be thinking that you could keep track of bluetooth connections to indicate if a bluetooth controller was attached or watch USB connections for USB devices. This is not a recommended approach for two main reasons.
+
+First of all, the information you are able to guess about connected devices based on API variables is not consistent and the number of bluetooth/hardware/stylus devices is continually growing.
+
+The second reason to use received events instead of connection status is that users may have a mouse, bluetooth controller, MIDI controller, etc. connected but not be actively using it to interact with your app or game.
+
+By responding to input events that have been actively received by your app, you ensure you are responding to your users real actions in real-time, and not trying to guess their intentions with incomplete information.
+
+### Example: game with touch, keyboard/mouse, and controller support
+
+Imagine you have developed a car racing game for touch-based mobile phones. Players can accelerate, decelerate, turn right, turn left, or use nitro for a speed boost.
+
+The current touchscreen interface consists of an on-screen joystick on the bottom left of the screen for the speed and direction controls, and a button on the bottom right for nitro. The user can collect nitro canisters on the track and when the nitro bar at the bottom of the screen is full, a message appears above the button that says “Press for Nitro!”. If it’s the user’s first game or no input is received for a while, “tutorial” text appears above the joystick showing the user how to make the car move.
+
+You would like to add keyboard and bluetooth game controller support. Where do you begin?
+
+![Car racing game with touch controls](/images/games/optimizing-games-input/input-touch.png)
+
+#### Input states
+
+Begin by identifying all of the **input states** your game might be running in and then listing all the parameters that you would like to change in each state.
+
+<table>
```

Comment:
Please write this table [using Markdown](https://www.markdownguide.org/extended-syntax/). You can use this [table generator](https://www.tablesgenerator.com/markdown_tables) as a starting place if you'd like